### PR TITLE
Add TimeKeeper class #6051

### LINKED ISSUE #6051: Add TimeKeeper class
The `TimeKeeper` class will be used to manually throw `DeadlineExceededException` earlier than the Google App Engine default (at 1 second left), to allow more time for the code to catch and handle the exception.

### PR DESCRIPTION
**Pull request by samsontmr**
Fixes #6051 

Continuing from #6052 

[WIP]
TODO: Class header coments

### CODE REVIEW

File: src/main/java/teammates/common/util/RequestTimeKeeper.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    private final long remainingMillisThreshold;
+    private long deadlineExpiryTime;
+
+    public RequestTimeKeeper(long remainingMillisThreshold) {
+        this.remainingMillisThreshold = remainingMillisThreshold;
+        deadlineExpiryTime = getCurrentTimeInMillis() + ApiProxy.getCurrentEnvironment().getRemainingMillis();
+    }
+
+    private long getCurrentTimeInMillis() {
+        return new Date().getTime();
```
Comment (Neurrone) [inline src/main/java/teammates/common/util/RequestTimeKeeper.java]: TimeHelper.getMsOffsetToCurrentTime could be used here.

File: src/main/java/teammates/common/util/RequestTimeKeeper.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    public void confirmEnoughTimeLeft() throws DeadlineExceededException {
+        if (!hasEnoughTime()) {
+            throw new DeadlineExceededException();
+        }
+    }
+    
+    public boolean hasEnoughTime() {
+        return deadlineExpiryTime - getCurrentTimeInMillis() > remainingMillisThreshold;
```
Comment (Neurrone) [inline src/main/java/teammates/common/util/RequestTimeKeeper.java]: And here too.

**Comment by wkurniawan07** [comment]
@samsontmr let the reviewers know when this is ready for review. Just one small thing from me, you should not forget to also define where this class is going to be called, otherwise you are introducing an essentially unused class.

**Comment by wkurniawan07** [comment]
@Neurrone since this PR has a more concrete direction, it can proceed first. #6300 does not have clear goal yet as of now.

**Comment by samsontmr** [comment]
@wkurniawan07 got it! I temporarily removed the part where it was called due to a bunch of merge conflicts

**Comment by samsontmr** [comment]
Ready for review! @chowyb @Neurrone

File: src/main/java/teammates/common/util/RequestTimeKeeper.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+     * * Checks if enough time is remaining and throws a DeadlineExceededException if there is not.
+     * @throws DeadlineExceededException
+     */
+    public void confirmEnoughTimeLeft() throws DeadlineExceededException {
+        if (!hasEnoughTime()) {
+            throw new DeadlineExceededException();
+        }
+    }
+    
+    public boolean hasEnoughTime() {
```
Comment (Neurrone) [inline src/main/java/teammates/common/util/RequestTimeKeeper.java]: Is this method intended for public usage? If not, suggest making it private.

Comment (samsontmr) [inline src/main/java/teammates/common/util/RequestTimeKeeper.java]: Yup it's supposed to be a public method.

**Comment by wkurniawan07** [comment]
I was afraid something like this would happen. Looks like Codecov still needs more tweaks. Sorry that this PR gets a collateral damage 😅
You can disregard the red Xs caused by coverage reports for now, although it'll be ideal if you can turn those red Xs to green ticks 😛

**Comment by chowyb** [comment]
Just to add on here since the description of the issue might be a bit vague:

This issue was originally created since at the time we were receiving a larger than usual amount of `HardDeadlineExceededError` from the longer methods, which is thrown after 60 seconds. The default `DeadlineExceededException` is thrown after 59 seconds, to allow sufficient time to clean up and fail gracefully, but in some cases the 1 second provided was not sufficient. Therefore, the idea was to have a TimeKeeper class to manually throw the `DeadlineExceededException` earlier than the default one, to allow more time for the cleanup.

`getFeedbackSessionResultsForUserWithParams` was identified as a problematic method that could benefit from the use of such a class due to its tendency to exceed the deadline (at the time anyway, not sure if this issue still persists). Therefore, for the initial introduction of this class, it was intended to be used there to see how effective it is at reducing the occurrence rate of `HardDeadlineExceededError`.

However, from what @wkurniawan07 has said, this might not work for unit tests, yet it seems unlikely to work effectively outside of the method itself (at least in this particular case, since `*Action` classes usually only call this method once, making it unlikely that the deadline is exceeded while control is within the `*Action` class).

**Comment by samsontmr** [comment]
@chowyb would you suggest leaving it in `getFeedbackSessionResultsForUserWithParams`? Perhaps there could be a check to ensure a GAE environment exists, and if a GAE environment is not found, `RequestTimekeeper` would use a pre-specified/user specified duration?

**Comment by wkurniawan07** [comment]
> it seems unlikely to work effectively outside of the method itself (at least in this particular case, since *Action classes usually only call this method once, making it unlikely that the deadline is exceeded while control is within the *Action class)

If I understand this correctly, the scenario will be like:
```java
@Override
protected void execute() {
    ... // 30 seconds elapsed here
    legendaryMethod(); // 50 seconds elapsed here (e.g.)
    ...
}
```
This tells one thing for sure: the problem exists in the legendary method itself (20 seconds for a method to finish???). #6576 just found itself a new use.

**Comment by wkurniawan07** [comment]
@samsontmr are you going to continue on this one?

**Comment by samsontmr** [comment]
Yep, it has to wait for the legendary method to be refactored first before continuing right?

**Comment by samsontmr** [comment]
Now that the legendary method has been refactored, shall I proceed with this? Or has this already become obsolete :joy:

**Comment by damithc** [comment]
@chowyb and I thought of adding a TimeKeeper class to measure time taken by a request and terminate the request a few seconds before it reaches the max time limit, to give us enough time to clean up. 
However, I recently noticed that GAE API has a method that can tell us the elapsed time of a request. So we might not need a TimeKeeper class at all (at least not the way we initially conceptualized it)

**Comment by samsontmr** [comment]
@damithc Shall I close this PR then?

**Comment by damithc** [comment]
> Shall I close this PR then?

Sure. Alternatively, you can try using the GAE API method instead of creating our own timekeeper.

**Comment by wkurniawan07** [comment]
Let's close this until we have a more concrete plan on what the issue is going to be.

**Pull request closed without merging**